File: common/src/main/java/maestro/api/pathing/goals/GoalBlock.java
package maestro.api.pathing.goals;

import maestro.api.utils.BetterBlockPos;
import maestro.api.utils.SettingsUtil;
import maestro.api.utils.interfaces.IGoalRenderPos;
import net.minecraft.core.BlockPos;

/** A specific BlockPos goal */
public class GoalBlock implements Goal, IGoalRenderPos {

    /** The X block position of this goal */
    public final int x;

    /** The Y block position of this goal */
    public final int y;

    /** The Z block position of this goal */
    public final int z;

    public GoalBlock(BlockPos pos) {
        this(pos.getX(), pos.getY(), pos.getZ());
    }

    public GoalBlock(int x, int y, int z) {
        this.x = x;
        this.y = y;
        this.z = z;
    }

    @Override
    public boolean isInGoal(int x, int y, int z) {
        return x == this.x && y == this.y && z == this.z;
    }

    @Override
    public double heuristic(int x, int y, int z) {
        int xDiff = x - this.x;
        int yDiff = y - this.y;
        int zDiff = z - this.z;
        return calculate(xDiff, yDiff, zDiff);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }

        GoalBlock goal = (GoalBlock) o;
        return x == goal.x && y == goal.y && z == goal.z;
    }

    @Override
    public int hashCode() {
        return (int) BetterBlockPos.longHash(x, y, z) * 905165533;
    }

    @Override
    public String toString() {
        return String.format(
                "GoalBlock{x=%s,y=%s,z=%s}",
                SettingsUtil.maybeCensor(x),
                SettingsUtil.maybeCensor(y),
                SettingsUtil.maybeCensor(z));
    }

    /**
     * @return The position of this goal as a {@link BlockPos}
     */
    @Override
    public BlockPos getGoalPos() {
        return new BlockPos(x, y, z);
    }

    public static double calculate(double xDiff, int yDiff, double zDiff) {
        double heuristic = 0;

        // Vertical heuristic: Use swimming cost (3.5) as minimum estimate
        // This provides a better heuristic for underwater goals than terrestrial movement costs
        // TODO: Ideally this would detect if we're actually in water, but using the minimum
        // of swimming vs terrestrial costs ensures the heuristic is admissible (never
        // overestimates)
        double verticalHeuristic = GoalYLevel.calculate(0, yDiff);
        double swimmingVerticalHeuristic = Math.abs(yDiff) * 3.5; // SWIM_UP/DOWN cost
        heuristic += Math.min(verticalHeuristic, swimmingVerticalHeuristic);

        // use the pythagorean and manhattan mixture from GoalXZ
        heuristic += GoalXZ.calculate(xDiff, zDiff);
        return heuristic;
    }
}

File: common/src/main/java/maestro/api/pathing/goals/GoalBlock.kt
package maestro.api.pathing.goals

import maestro.api.utils.SettingsUtil
import maestro.api.utils.interfaces.IGoalRenderPos
import maestro.api.utils.pack
import net.minecraft.core.BlockPos
import kotlin.math.abs
import kotlin.math.min

/** A specific BlockPos goal */
open class GoalBlock(
    @JvmField val x: Int,
    @JvmField val y: Int,
    @JvmField val z: Int,
) : Goal,
    IGoalRenderPos {
    constructor(pos: BlockPos) : this(pos.x, pos.y, pos.z)

    override fun isInGoal(
        x: Int,
        y: Int,
        z: Int,
    ): Boolean = x == this.x && y == this.y && z == this.z

    override fun heuristic(
        x: Int,
        y: Int,
        z: Int,
    ): Double {
        val xDiff = x - this.x
        val yDiff = y - this.y
        val zDiff = z - this.z
        return calculate(xDiff.toDouble(), yDiff, zDiff.toDouble())
    }

    override fun getGoalPos(): BlockPos = BlockPos(x, y, z)

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is GoalBlock) return false
        return x == other.x && y == other.y && z == other.z
    }

    override fun hashCode(): Int = (pack(x, y, z).packed * 905165533).toInt()

    override fun toString(): String =
        "GoalBlock{x=${SettingsUtil.maybeCensor(x)},y=${SettingsUtil.maybeCensor(y)},z=${SettingsUtil.maybeCensor(z)}}"

    companion object {
        @JvmStatic
        fun calculate(
            xDiff: Double,
            yDiff: Int,
            zDiff: Double,
        ): Double {
            var heuristic = 0.0

            // Vertical heuristic: Use swimming cost (3.5) as minimum estimate
            // This provides a better heuristic for underwater goals than terrestrial movement costs
            // TODO: Ideally this would detect if we're actually in water, but using the minimum
            // of swimming vs terrestrial costs ensures the heuristic is admissible (never overestimates)
            val verticalHeuristic = GoalYLevel.calculate(0, yDiff)
            val swimmingVerticalHeuristic = abs(yDiff) * 3.5 // SWIM_UP/DOWN cost
            heuristic += min(verticalHeuristic, swimmingVerticalHeuristic)

            // use the pythagorean and manhattan mixture from GoalXZ
            heuristic += GoalXZ.calculate(xDiff, zDiff)
            return heuristic
        }
    }
}
